WIP: Use normal hash operator classes for data distribution.

hlinnaka opened the pull request:
Replace the use of the built-in hashing support for built-in datatypes,
in cdbhash.c, with the normal PostgreSQL hash functions.

It is no longer enough to track which columns/expressions are used to
distribute data. You also need to know the hash function used. For that,
a new field is added to gp_distribution_policy, to record the hash
operator class used for each distribution key column. In the planner,
a new opfamily field is added to DistributionKey, to track that throughout
the planning.

Normally, if you do "CREATE TABLE ... DISTRIBUTED BY (column)", the default
hash operator class for the datatype is used. But this patch extends the
syntax so that you can specify the operator class explicitly, like
"... DISTRIBUTED BY (column opclass)". This is similar to how an operator
class can be specified for each column in CREATE INDEX.

To support upgrade, the old hash functions have been converted to special
(non-default) operator classes, named cdbhash_*_ops. For example,
if you want to use the old hash function for an integer column, you could
do "DISTRIBUTED BY (intcol cdbhash_int4_ops)". The old hard-coded whitelist
of operators that have "compatible" cdbhash functions has been replaced
by putting the compatible hash opclasses in the same operator family.
For example, all legacy integer operator classes, cdbhash_int2_ops,
cdbhash_int4_ops and cdbhash_int8_ops, are all part of the
cdbhash_integer_ops operator family).

This removes the pg_database.hashmethod field. The hash method is now
tracked on a per-table and per-column basis, using the opclasses, so it's
not needed anymore.

ORCA doesn't know about hash operator classes, or the possibility that we
might need to use a different hash function for two columns with the same
datatype. Therefore, it cannot produce correct plans for queries that mix
different distribution hash opclasses for the same datatype, in the same
query. There are checks in the DXL->Query translation, to detect that case,
and fall back to planner. As long as you stick to the default opclasses in
all tables, we let ORCA to create the plan without any regard to them, and
use the default opclasses when translating the DXL plan to a Plan tree.
We also allow the case that all tables in the query use the "legacy"
opclasses, so that ORCA can also be used after pg_upgrade. But a mix of
the two, or using any non-default opclasses, forces ORCA to fall back.

One curiosity with this is the "int2vector" and "aclitem" datatypes. They
have a hash opclass, but no b-tree operators. GPDB 4 used to allow them
as DISTRIBUTED BY columsn, but we forbid that in GPDB 5, in commit
56e7c16bad. Now they are allowed again, so you can specify an int2vector
or aclitem column in DISTRIBUTED BY, but it's still pretty useless,
because the planner still can't form EquivalenceClasses on it, and will
treat it as "strewn" distribution, and won't co-locate joins.

TODOs:

* There are no legacy anyarray/anyrange opclasses. Because the cdbhash
  code for them was kind of broken.
  (see https://github.com/greenplum-db/gpdb/issues/5467).

* A few FIXME comments in the code.

Discussion: https://groups.google.com/a/greenplum.org/forum/#!topic/gpdb-dev/4fZVeOpXllQ

hlinnaka commented:
This is in pretty good shape now. The last two commits are the ones you should focus on as part of this PR ("Use normal hash operator classes for data distribution" and "Rename gp_distribution_policy.attrnums to distkey, and make it int2vector". The rest of the commits are discussed separately, in other PRs, but this work depends on them, so I had to include them in this PR, too.

I'm awaiting review on a couple of other PRs that this depends on:

* PR https://github.com/greenplum-db/gpdb/pull/6253
* PR https://github.com/greenplum-db/gpdb/pull/6314

There are some known regression failures, caused by changed plans, because this PR changes the hash functions and hence the distribution of data. 

There are also a few "non stream bitmap" failures, caused by issue https://github.com/greenplum-db/gpdb/issues/5353, and one failure caused by https://github.com/greenplum-db/gpdb/issues/6074.

mkiyama commented:
@hlinnaka  couple of questions.
--Will this WIP be changed to a "regular" PR that will be merged?
--Tables defined with a non-default hash will fall back to the planner.  Is ORCA support planned for 6?

hlinnaka commented:
> @hlinnaka couple of questions.
> --Will this WIP be changed to a "regular" PR that will be merged?

Not sure. I guess I'll just remove the "WIP" tag, after this is truly ready to be merged. Or if this requires some major overhauling based on review commits, then I might create a new PR.

> --Tables defined with a non-default hash will fall back to the planner. Is ORCA support planned for 6?

I don't know. I'm not planning to work any more on that, but I don't know what plans the other teams have.

mkiyama commented:
ok, I'll look through the doc changes.

mkiyama commented:
@hlinnaka 
I have some fixes changes for the dita docs. 
Most of them are fixes to make the XML valid.
I have some edits/changes for the example in ddl-table.xml

How do you wan to handle the changes?

hlinnaka commented:
On 27/11/2018 23:32, Mel Kiyama wrote:
> @hlinnaka <https://github.com/hlinnaka>
> I have some fixes changes for the dita docs.
> Most of them are fixes to make the XML valid.
> I have some edits/changes for the example in ddl-table.xml
> 
> How do you wan to handle the changes?

Thanks! Feel free to push them as an additional commit to this PR, and 
I'll squash it with the main commit later. Or post the diff or the 
modified files here as a comment, or whatever is most convenient for 
you, really. I'll figure out how to incorporate the changes.

- Heikki

mkiyama commented:
@hlinnaka 
One other thing, I'm kind of confused by the query and EXPLAIN at the end of the example in in ddl-table.xml

The EXPLAIN plan has redstribute motion nodes, but the text description implies that the plan should not have them.

And I could not get an EXPLAIN plan with redistribute motions using a really small 1 host/3segment GPDB 6.

bhuvnesh2703 commented on src/backend/gpopt/translate/CContextDXLToPlStmt.cpp:
```diff
@@ -255,6 +260,85 @@ CContextDXLToPlStmt::AddRTE
 		m_result_relation_index = gpdb::ListLength(*(m_rtable_entries_list));
 	}
+
+	// What opclasses are being used in the distribution policy?
```
Question: Any specific reason to have the fallback mechanism in DXL to PlStmt instead of CTranslatorQuerytoDXL, can we iterate over the from list in tree to do similar checks? Since DXLToPlStmt is the last step,  if we can fallback way early it should be better.

bhuvnesh2703 replied:
also what about partition tables if the root has default op class and any of the leaf partition has legacy? is that possible even?

bhuvnesh2703 replied:
i checked on this branch, the cases which you mentioned-
1. Query has all tables with legacy => Works
2. Query has all tables with default => Works
3. Query has mixed legacy and default => Fallsback

hlinnaka replied:
> Question: Any specific reason to have the fallback mechanism in DXL to PlStmt instead of CTranslatorQuerytoDXL (or even RelcacheToDXL), can we iterate over the from list in tree to do similar checks? Since DXLToPlStmt is the last step, if we can fallback way early it should be better.

Makes sense. I rewrote it that way.

One corner-case is that the original query might reference tables that are completely stripped away from the final plan. We wouldn't need to take those into account in deciding which opclasses to use, but now we do. But that's acceptable.

bhuvnesh2703 replied:
Thanks @hlinnaka .. I reviewed the orca changes.

Except the partition table case, where one of the partition is using a default op class vs other which are using legacy, it doesn't fallback currently. I think we should.

Other than that, LGTM. Do you plan to handle the partition table case?

mkiyama commented:
@hlinnaka 
Here is a file with my suggested edits for the docs, I hope that's what it is anyway.

Let me know if you have any questions.

FYI - there was a note about the operator classes and upgrading to 6.0. I moved that information to the 6.0 Release Notes.

[opclass-docs-patch.zip](https://github.com/greenplum-db/gpdb/files/2626606/opclass-docs-patch.zip)

kainwen commented:
Hi @hlinnaka 
Thanks for this PR.

I am still reviewing this PR. IIUC, it removes the restriction that all tables in a database have to use the same cdbreducehash method, now it is determined by opclass of each column, which sort of eliminate the job of `migrate a database from old modulo hash to new jump hash` because we could finish the same obj via `expand` (at least CTAS can do it).

Am I right?

hlinnaka commented:
On 4 December 2018 05:38:32 EET, Zhenghua Lyu <notifications@github.com> wrote:
>Hi @hlinnaka 
>Thanks for this PR.
>
>I am still reviewing this PR. IIUC, it removes the restriction that all
>tables in a database have to use the same cdbreducehash method, now it
>is determined by opclass of each column, which sort of eliminate the
>job of `migrate a database from old modulo hash to new jump hash`
>because we could finish the same obj via `expand` (at least CTAS can do
>it).
>
>Am I right?

Yep, that's right.

- Heikki

gaos1 commented on src/backend/commands/dbcommands.c:
```diff
@@ -523,8 +523,6 @@ createdb(const CreatedbStmt *stmt)
 	new_record[Anum_pg_database_datminmxid - 1] = TransactionIdGetDatum(src_minmxid);
 	new_record[Anum_pg_database_dattablespace - 1] = ObjectIdGetDatum(dst_deftablespace);
-	/* new created dbs all use jump hash */
-	new_record[Anum_pg_database_hashmethod - 1] = Int32GetDatum(JUMP_HASH_METHOD);
```
Without this field, I guess all new created table will use new operator with jump consistent hash, while other tables in the same database will still use operator with legacy modulo after upgrade.  Even a join query on an old table and new table can work, it will need motion and inefficient. 

We are thinking if it is more reasonable to postpone the data migration and change them to be jump consistent compatible until next expansion of the cluster.  If users don't plan to expand their clusters in short term, they cannot get extra benefits with reshuffle of data after upgrade 6. 

it might be better and simple to keep this field, and let pg/gp_upgrade mark old databases to use legacy modulo. so all tables in this database are using the same legacy modulo and its operator, including new created tables after upgrade. For all new created databases, we can just move to use jump consistent hash.

hlinnaka replied:
Ok. So if we keep the per-database flag, then the behavior should be that if the per-database flag says "legacy", then CREATE TABLE will choose the use the legacy hash opclasses by default, if available.

hlinnaka replied:
Addressed this.

I did not restore the pg_database field. However, I added a new GUC, `gp_use_legacy_hashops` for this. That's more flexible to use in testing, than a whole new pg_database field. When the GUC is set, and you CREATE TABLE withoutt specifying an opclass explicitly, the legacy opclasses are used. T

The way this now works, is that pg_upgrade sets the `gp_use_legacy_hashops` GUC for as a database-level setting, for each upgraded database. So after upgrade, the legacy opclasses are used for any newly created tables too. But you can change the default easily by clearing the GUC with ALTER TABLE. The to-be-written tool that converts all tables to the new opclasses after upgrade, is expected to clear the GUC.

The GUC also affects how the opclasses are printed in `\d <table>` and in pg_dump. When the GUC is set, legacy opclasses are treated as the default, and are not explicitly printed. So if you pg_dump an upgraded database, and restore into a new database, the new default opclasses will be used in the new database.

yydzero replied:
+1 for GUC, thus we don't need to introduce 'CREATE DATABASE dbname WITH (distribution_hash_method=modulo)' anymore. @gaos1 @kainwen @pivotal-ning-yu

yydzero commented:
@pivotal-ning-yu @xiong-gang @pengzhout @sshang would you please help to review this? Thanks!

pivotal-ning-yu commented on gpMgmt/bin/gpload.py:
```diff
@@ -2653,6 +2653,9 @@ def get_qualified_tablename(self):
     def get_table_dist_key(self):
+        # FIXME: refers to attrnums, not distkey, so this query works on GPDB5 but
+        # not on GPDB6. How to do version checks here? Does this need to work against
+        # old servers?
```
resgroup has version check like below, maybe we can do similar check here.

https://github.com/greenplum-db/gpdb/blob/57c599bc69a25fd07d4b590aba0e8975632951cd/gpMgmt/bin/gpcheckresgroupimpl#L9-L12

https://github.com/greenplum-db/gpdb/blob/57c599bc69a25fd07d4b590aba0e8975632951cd/gpMgmt/bin/gpcheckresgroupimpl#L68

yydzero replied:
@adam8157 Would you please help to take a look? Thanks!

adam8157 replied:
gptransfer doesn't need to work with different version servers, because the gppylib doesn't support.

gppylib has a method GpVersion() parsing the `SELECT version()` https://github.com/greenplum-db/gpdb/blob/7148fc463345af67da770c064a4fa25444ddbd9e/gpMgmt/bin/gptransfer#L4457

Go utilities do the same thing, https://github.com/greenplum-db/gp-common-go-libs/blob/459655a5c915e633b3a15987ae469ab64b3405a4/dbconn/version.go#L33

pf-qiu replied:
Hi, @hlinnaka . gpload will be bound to a specific gpdb version, so no need to check that.
Actually the gpload script for 4 and 5 differs due to error tables. We can do this for 6 also.
I assume the `distKey` column is just renamed from `attrnums` column for 5, and they are the same thing.
So simply rename the column in that sql of gpload should be fine. 
I've run all tests on 6, and test the update error, it functions correctly.

pivotal-ning-yu commented:
The other 4 commits also looks good to me in general, but I vote for keeping the per-database reduce method attribute.

melanieplageman commented on src/backend/cdb/cdbcat.c:
```diff
@@ -292,59 +321,51 @@ GpPolicyFetch(Oid tbloid)
 	 */
 	if (HeapTupleIsValid(gp_policy_tuple))
 	{
+		Form_gp_policy policyform = (Form_gp_policy) GETSTRUCT(gp_policy_tuple);
```
Why are we able to use `GETSTRUCT` here? Is it because this is a system tuple with fixed size attributes?

hlinnaka replied:
That's right. I reordered the columns so that all the fixed-size attributes are first. That way, they can be accessed directly through the struct.

hlinnaka commented:
Rebased, and re-wrote the ORCA translation code to decide which opclasses to use per Bhuvnesh's suggestion.

@melanieplageman, you mentioned offlist that you were getting "cache lookup failed" errors with this. I found the cause: I fumbled an earlier rebasing, in a way that broke simple INSERT commands with ORCA. Should work now.

yydzero commented on src/backend/utils/misc/guc_gp.c:
```diff
@@ -622,6 +623,17 @@ struct config_bool ConfigureNamesBool_gp[] =
 
+	{
+		{"gp_use_legacy_hashops", PGC_USERSET, COMPAT_OPTIONS_PREVIOUS,
+			gettext_noop("If set, new tables will use legacy distribution hashops by default"),
```
is it better to use SUSET to make it super user only?

hlinnaka replied:
That would be rather pointless. The user can specify any operator class explicitly in the CREATE TABLE command, this GUC just changes the default one, if you don't specify one explicitly.

pivotal-ning-yu commented:
On Thu, Dec 20, 2018 at 5:28 PM Heikki Linnakangas <notifications@github.com>
wrote:

> *@hlinnaka* commented on this pull request.
> ------------------------------
>
> In gpMgmt/bin/gpload.py
> <https://github.com/greenplum-db/gpdb/pull/6327#discussion_r243203633>:
>
> > @@ -2653,6 +2653,9 @@ def get_qualified_tablename(self):
>          return tblname
>
>      def get_table_dist_key(self):
> +        # FIXME: refers to attrnums, not distkey, so this query works on GPDB5 but
> +        # not on GPDB6. How to do version checks here? Does this need to work against
> +        # old servers?
>
> Hmm, that version check gets the version number from the installed
> binaries. We need to check the version of the database that you're
> connected to. I don't think those are necessarily the same in gpload and
> gptransfer.
>
> (I'm surprised that gpcheckresgroupimpl does it that way either. Is that
> tool really expected to be used against a cluster on a different version?)
>

gpcheckresgroupimpl is invoked by gpconfig, not sure if it can be used on a
different version.  It was suggested by Jimmy at PR #4821,
https://github.com/greenplum-db/gpdb/pull/4821#discussion_r179220867 .


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/6327#discussion_r243203633>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AYjBHlDGIBbA12Fz96AcYsOGewvIm-lnks5u61hKgaJpZM4YzZl2>
> .
>

hlinnaka commented:
On 21/12/2018 23:37, Bhuvnesh wrote:
> Except the partition table case, where one of the partition is using a 
> default op class vs other which are using legacy, it doesn't fallback 
> currently. I think we should.

We shouldn't let that scenario to occur in the first place, I think. If 
two partitions are using different opclasses, it's the same as using 
different distribution policies altogether. I'll try to add a check for 
that in ALTER TABLE.

- Heikki

bhuvnesh2703 commented:
> We shouldn't let that scenario to occur in the first place, I think. If
two partitions are using different opclasses, it's the same as using
different distribution policies altogether. I'll try to add a check for
that in ALTER TABLE.

Thanks, LGTM.

hlinnaka commented:
> On 21/12/2018 23:37, Bhuvnesh wrote: Except the partition table case, where one of the partition is using a default op class vs other which are using legacy, it doesn't fallback currently. I think we should.
> We shouldn't let that scenario to occur in the first place, I think. If two partitions are using different opclasses, it's the same as using different distribution policies altogether. I'll try to add a check for that in ALTER TABLE

Now that I test that, I think this patch already checks for that correctly:

```
postgres=# create table parttab (x int4) distributed by (x) partition by range (x) (start (1) end (100) every (50)) ;
NOTICE:  CREATE TABLE will create partition "parttab_1_prt_1" for table "parttab"
NOTICE:  CREATE TABLE will create partition "parttab_1_prt_2" for table "parttab"
CREATE TABLE
postgres=# create table xtab (x int4) distributed by (x cdbhash_int4_ops);
CREATE TABLE
postgres=# alter table parttab exchange partition for (2) with table xtab;
ERROR:  distribution policy for "xtab" must be the same as that for "parttab"
```

d commented:
Hey @hlinnaka I am locally running reg test and occasionally I'm getting an error like `ERROR:  cache lookup failed for opclass 0 (lsyscache.c:1183)` during `tinterval`. Did this happen to you? I cannot reliably reproduce this error though, which scares me more.

d commented:
Looks like I am not alone -- luckily -- the PR pipeline also hit this error in https://prod.ci.gpdb.pivotal.io/builds/189627

hlinnaka commented:
On 02/01/2019 19:44, Jesse Zhang wrote:
> Hey @hlinnaka I am locally running reg test and occasionally I'm
> getting an error like `ERROR:  cache lookup failed for opclass 0
> (lsyscache.c:1183)` during `tinterval`.

I saw that in the PR pipeline, too. I believe it is because we don't 
have a hash opclass for 'tinterval' (see 
https://groups.google.com/a/greenplum.org/forum/#!topic/gpdb-dev/x7YNr1cDbos). 
  Although, you shouldn't get a user-unfriendly error like that, even if 
a hash opclass is missing, so we probably need to work on the checks in 
orca translator.

> Did this happen to you? I cannot reliably reproduce this error
> though, which scares me more.
Oh, ok, that makes me a bit scared too. It always happens for me, in the 
same places as in the PR pipeline.

- Heikki

d commented:
I wrote:

> Hey @hlinnaka I am locally running reg test and occasionally I'm getting an error like `ERROR: cache lookup failed for opclass 0 (lsyscache.c:1183)` during `tinterval`. Did this happen to you? I cannot reliably reproduce this error though, which scares me more.

Heikki wrote:
> Oh, ok, that makes me a bit scared too.

False alarm! I checked again on my "passing" run, and turns out I was running against 5.10 ... My bad...

d commented:
Big question: given that we are supporting explicit operator class, should we also support distributing on an expression?

Context:
It's a little awkward that both `bpchar` and `text` are in the same hash operator family:

1. I understand that when creating a (hypothetical) hash index, a `bpchar` attribute will get the same hash values as a `text`, if they "compare equally". So semantically, `=(bpchar,bpchar)` and `=(text, text)` should be in the same op family.
1. However, `bpchar` is not directly "comparable" to `text`, in the sense that there isn't a binary coercible cast from one to the other (planner has a more complete and convoluted logic, but we'll grossly simplify it here). That means that planner might need significant work to ever make use of the fact that we _do_ co-locate `bpchar` and `text`.
1. Upstream Postgres does not ship built-in hash index support for text types. But if anyone ever does supply that, I'd imagine we only do it for `text` type. Users who want to create a hash index on a `bpchar` column `b` of table `foo(a int, b char(64))` can simply index on the expression `b::text`.

This is *not* a regression, to be clear, because we seem to always redistribute when it comes to joining on a `=(bpchar, text)` operator. But it feels like we're so close to have a good abstraction to enable it.

d commented:
I've been thinking hard about the `tinterval` test case.

The logic for ORCA translator is a little scrappy:
1. We examine all the distribution keys for all tables involved in the query, check for "mixed" and "non-default" situations and bail.
1. ORCA remains oblivious of the distinction between different op classes, goes its merry way and generates a plan.
1. The translator uses what it remembers from the first step (what did I see in the query?) and "guesses" the op class for each motion. But we _pick default_ if we don't recall anything.

I think it's wise to resist the urge to "guess harder" in step three to pick a hash op class for redistribute motions. Instead, we should inspect the query more in step 1:

1. We should continue to inspect all distribution keys; in addition
1. We should also check the "=" operators used in join conditions;
1. We should also check the implied (equal) operators for `GROUP BY`, `DISTINCT`, and possibly `DISTINCT ON`.

hlinnaka commented:
On 04/01/2019 00:54, Jesse Zhang wrote:
> I've been thinking hard about the `tinterval` test case.
> 
> The logic for ORCA translator is a little scrappy:
> 1. We examine all the distribution keys for all tables involved in the query, check for "mixed" and "non-default" situations and bail.
> 2. ORCA remains oblivious of the distinction between different op classes, goes its merry way and generates a plan.
> 3. The translator uses what it remembers from the first step (what did I see in the query?) and "guesses" the op class for each motion. But we _pick default_ if we don't recall anything.
> 
> I think it's wise to resist the urge to "guess harder" in step three to pick a hash op class for redistribute motions. Instead, we should inspect the query more in step 1:
> 
> 1. We should continue to inspect all distribution keys; in addition
> 2. We should also check the "=" operators used in join conditions;
> 3. We should also check the implied (equal) operators for `GROUP BY`, `DISTINCT`, and possibly `DISTINCT ON`.

Hmm. What would you check in the "=" operators, or in GROUP BY or DISTINCT?

Looking at that 'tinterval' case under gdb, it looks like ORCA tries to 
create a plan with a Motion node, hashed on the 'tinterval' column. Why 
does it do that? Ah, the "=" operator for tinterval is marked as 
'oprcanhash=true', even though it doesn't have a default hash operator 
class. I forget if that's appropriate; is oprcanhash supposed to be set, 
if the datatype has *any* hash operator class, or only if it has a 
default? If it's for *any* hash opclass, then we'll need to change the 
ORCA translator to not look at that, but look at the presence of a 
default hash opclass, instead.

- Heikki

d commented:
I don't think orca considers `oprcanhash` at all, and it always checks all op classes of the `=` operator to find whether there's at least one hash operator class, in order to generate a hash join

hlinnaka commented:
I added code to pg_upgrade to check for 'abstime', 'reltime', or 'tinterval' datatypes being used in distribution key columns. However, I'm not sure how to test it. Do we have test scripts somewhere to test upgrading from GPDB 5 to 6?

hlinnaka commented:
The regression failure in the 'partition' test turned out to be an existing bug in COPY FROM. It was exposed by the existing test, now that the data distribution happens to be different. Opened issue #6678 for that.

hlinnaka commented:
Status update:

All the per-requisite issues that this PR depended on, have now been fixed. The regression tests without ORCA are passing.

There are a couple of TODOs left:
* pg_upgrade should check for array distribution key columns, and throw an error if there are any. They never worked correctly in GPDB 5, and hence there is no legacy opclass support for them. Also, the new pg_upgrade checks that this adds, need to be tested.
* There is a crash with ORCA in the CI. I could not reproduce that on my laptop. (Although I got some other assertion failures with Debug-enabled ORCA from the same tests, which didn't help.)
* There are "cache lookup failed for opclass 0" errors with ORCA, the ones we discussed with @d.
* gptransfer and gpload need to be modified for the changes in gp_distribution_policy catalog table. I added FIXMEs comments about that. @adam8157, can fix that, please?

I need help with these remaining items. Any volunteers to work on the pg_upgrade changes? Any volunteers to debug the ORCA issues?

xm2001 commented:
mpp team would like to help on the upgrade todo and work out a commit for review.

gaos1 commented:
mpp team would like to work out a commit on pg_upgrade change for review.

yydzero commented:
@BrianPivotal @adam8157 Would be great if help the gptransfer and gpload fix.

adam8157 commented:
No problem, I will take the gptransfer and gpload.

adam8157 commented:
@lij55 @pf-qiu I'm not sure if the gpload's get_table_dist_key() could be worked around, please take it over? Thanks.

pengzhout commented:
> * pg_upgrade should check for array distribution key columns, and throw an error if there are any. They never worked correctly in GPDB 5, and hence there is no legacy opclass support for them. Also, the new pg_upgrade checks that this adds, need to be tested.

Hi @hlinnaka, I wanna to double check the "array distribution key" here, are we talking about is the int2vector[] and aclitem[] that need be checked and thrown an error because we don't have legacy opclass for them? For anyarray type like int[3][3], we have legacy hash opclass and default btree opclass for it, so I think "array distribution key" didn't mean all array type.

vraghavan78 commented:
@hardikar @bhuvnesh2703 can one of you look at the ORCA issue?

bhuvnesh2703 commented:
@hlinnaka looking into the CI Crash. Running the pipeline to see what is crashing.

asubramanya commented:
@hlinnaka We looked into the cache failures. It seems that for certain data types like `tinterval, abstime, money` there is no default hash opclass entry. If there is a redistribute motion on these types, then trying to lookup the default hash opclass will cause the cache lookup failures. We plan to fallback to planner if we are unable to find the hash opclass for a particular type during translation.

We are still looking into the wrong results issue.

CH + AS

pengzhout commented:
Pushed a commit to add a check to pg_upgrade for any distribution key, I tried to run 5->6 pg_upgrade manually and pg_upgrade failed before actually checking the distribution key type.

hlinnaka commented:
Thanks @pengzhout, @adam8157 and @pf-qiu for the fixes!

I rebased and force-pushed this, to kick the CI again. Waiting for the ORCA fixes.

hlinnaka commented:
@asubramanya fixed the ORCA issues at https://github.com/asubramanya/gpdb/tree/use-hash-opclass-for-distributed-by-pr, and I just cherry-picked the fixes here.

hlinnaka commented on src/test/regress/expected/gpdist_opclasses.out:
```diff
@@ -50,6 +50,8 @@ CREATE TABLE abstab_b (b int) DISTRIBUTED BY (b);
 -- The default opclass isn't helpful with the |=| operator, so this needs
 -- a Motion.
+-- Disable this test on ORCA for now
+set optimizer=off;
```
@asubramanya, did I understand correctly, that this was a pre-existing bug in ORCA? Is there a github issue for it?

asubramanya replied:
@hlinnaka Yes. This issue exists on 5X_STABLE as well. I created a story for it in our backlog. I'll create a github issue for this too.

asubramanya replied:
created github issue here: https://github.com/greenplum-db/gpdb/issues/6796

hlinnaka replied:
Thanks, I added a link to that issue in the comment, so that we remember to re-enable the test once that's fixed.

hlinnaka commented on src/test/regress/expected/gpdist_legacy_opclasses_optimizer.out:
```diff
@@ -261,7 +261,7 @@ explain (costs off) select * from legacy_enum a inner join legacy_enum b on a.co
  Gather Motion 3:1  (slice1; segments: 3)
    ->  Hash Join
-         Hash Cond: (legacy_enum.color = legacy_enum_1.color)
+         Hash Cond: (((legacy_enum.color)::anyenum = (legacy_enum_1.color)::anyenum) AND ((legacy_enum.color)::anyenum = (legacy_enum_1.color)::anyenum))
```
@asubramanya: That looks redundant. Is this a pre-existing ORCA issue, or something that happened in this PR?

hlinnaka commented:
Squashed, rebased and fixed the last minor issues discussed in the comments here. I'm opened a new PR for the final version, https://github.com/greenplum-db/gpdb/pull/6327. Closing this one.
